### PR TITLE
feat(playlists): implement playlist plugin

### PR DESCRIPTION
**Pull request by kuangp**
Signed-off-by: Phil Kuang <pkuang@factset.com>

## Hey, I just made a Pull Request!
(Apologies for the giant change here 😅)

I put together an implementation of my playlist idea from #6063 but generalized for all entities. Check out the READMEs and demo below for more details! 

https://user-images.githubusercontent.com/6998196/181414962-d7a03506-bc95-4a7d-8576-3740e6fde6d9.mov

### Potential Future Enhancements
- Clean up of stale entities refs (deleted entities that are still referenced by playlists)
- UI customization for integrators
- Playlist search collator / `ResultListItem` to add playlists to search
- More metadata to improve discoverability of useful playlists eg. tags?
- Pagination to support scaling 
- Distributed edits / conflict resolution support (if we want to get fancy 😁)
- Applying actions across entities in a playlist (eg. unregister all)
- Playlist plugins! (eg. Tech insights across entities in a playlist)

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
Hey @kuangp sorry I was holding off reviewing this as had a few other bits going on, will get to it this week! :pray:

File: .changeset/nice-ducks-smile.md
Comment (benjdlambert) [inline .changeset/nice-ducks-smile.md:2]: I'd make these `minor` changes

File: microsite/data/plugins/playlist.yaml
Comment (benjdlambert) [inline microsite/data/plugins/playlist.yaml:8]: Did you make this logo? Or where did it come from? Just need to make sure we can use it in the repo that's all

File: packages/core-components/api-report.md
Comment (benjdlambert) [inline packages/core-components/api-report.md:1366]: We probably want to do `Object.freeze` on https://github.com/backstage/backstage/blob/590442aac0a02e2271ca772a0c851a83f7422d32/packages/core-components/src/components/Table/Table.tsx#L57

Comment (benjdlambert) [reply packages/core-components/api-report.md:1366]: So that it looks nice in the report.

**Review by benjdlambert — COMMENTED** [review]
Just a quick pass right now, but I'm wondering about the naming of the plugin. `playlist` is a bit ambiguous. Not sure :shrug: @backstage/maintainers?

File: microsite/data/plugins/playlist.yaml
Comment (kuangp) [reply microsite/data/plugins/playlist.yaml:8]: hmm I found this on Google somewhere, don't remember where ... I can change this to something that's free to use, do you know if this logo is ok: https://www.flaticon.com/premium-icon/playlist_4081218?related_id=4081218&origin=tag

File: .changeset/nice-ducks-smile.md
Comment (kuangp) [reply .changeset/nice-ducks-smile.md]: updated

File: packages/core-components/api-report.md
Comment (kuangp) [reply packages/core-components/api-report.md]: done!

**Comment by kuangp** [comment]
> Just a quick pass right now, but I'm wondering about the naming of the plugin. `playlist` is a bit ambiguous. Not sure 🤷 @backstage/maintainers?

perhaps `entity-playlist` ?

**Comment by kuangp** [comment]
@benjdlambert @Rugvip quick reminder that this is still here awaiting further feedback 😁

**Comment by Rugvip** [comment]
@kuangp yep sorry, on my radar! It's a big one so needs a bigger chunk of time, which I haven't sliced out just yet 😅

**Comment by kuangp** [comment]
> @kuangp yep sorry, on my radar! It's a big one so needs a bigger chunk of time, which I haven't sliced out just yet 😅

no worries! just want to make sure this doesn't go stale 😄

File: plugins/playlist/src/components/PlaylistIndexPage/PlaylistIndexPage.tsx
Comment (Rugvip) [inline plugins/playlist/src/components/PlaylistIndexPage/PlaylistIndexPage.tsx:24]: I see that there's likely space for this filter layout system to be decoupled from the catalog and extracted into its own package :grin:

File: plugins/playlist/src/components/PlaylistPage/AddEntitiesDrawer.tsx
Comment (Rugvip) [inline plugins/playlist/src/components/PlaylistPage/AddEntitiesDrawer.tsx:133]: Oof, any thoughts on this @backstage/techdocs-core?

File: plugins/playlist/src/components/PlaylistPage/PlaylistEntitiesTable.tsx
Comment (Rugvip) [inline plugins/playlist/src/components/PlaylistPage/PlaylistEntitiesTable.tsx:74]: Another thing that might help is #12254. Could also be better (and faster) to move this to the backend? Think that should solve the catalog permission issue too.

File: plugins/search-react/src/components/SearchFilter/SearchFilter.tsx
Comment (Rugvip) [inline plugins/search-react/src/components/SearchFilter/SearchFilter.tsx:109]: @backstage/techdocs-core do these changes look alright?

**Review by Rugvip — CHANGES_REQUESTED** [review]
Holy sheeet, really solid work x)

Couple of concerns regarding the newly exposed APIs ofc 😅

About the name, I'm also feeling it's a little bit off, as in being too generic. Don't hate it though. Perhaps `entity-playlist` is a better option, although it's definitely not as short and neat. Maybe a couple more rounds of thinking while we iron out the rest?

File: packages/backend/src/plugins/permission.ts
Comment (Joonpark13) [reply packages/backend/src/plugins/permission.ts:42]: I don't see any issues with it myself! It's nice for plugin authors to offer a sensible default permission policy that integrators can just "plug and play" like this.

File: .changeset/nice-ducks-smile.md
Comment (kuangp) [reply .changeset/nice-ducks-smile.md:4]: sure, I'll add myself as a reviewer for the ADR plugin while I'm at it

File: plugins/playlist/src/components/PlaylistIndexPage/PlaylistIndexPage.tsx
Comment (kuangp) [reply plugins/playlist/src/components/PlaylistIndexPage/PlaylistIndexPage.tsx:24]: Totally agree ! along with the list provider hook stuff too which I mostly plagiarized 😅 
All that can definitely be moved out and made more generic to be reused in multiple plugins (but probably in a future change since this PR is large enough already 😁)

File: plugins/playlist-backend/migrations/20220701011329_init.js
Comment (kuangp) [reply plugins/playlist-backend/migrations/20220701011329_init.js:23]: hm I kinda avoided the suffix here since this gets unserialized into a `Playlist` object and since DBs don't support camel casing I didn't want to introduce complexity translating `ownerRef` to `owner_ref` and vice versa

File: plugins/playlist-backend/src/service/router.ts
Comment (kuangp) [reply plugins/playlist-backend/src/service/router.ts:126]: Went with your suggestion and moved the catalog entity reading to the backend to not expose the playlist entity refs to the frontend which should now respect catalog permissioning 🙂

**Comment by kuangp** [comment]
@Rugvip updated based on your feedback! let me know what you think 🙂

File: packages/core-components/api-report.md
Comment (Rugvip) [inline packages/core-components/api-report.md:474]: Realized this should obviously be `HeaderActionMenuItem` instead 😅 

Or `HeaderActionMenuActionItem`, but perhaps we don't need to go quite that far. Then again if we don't, perhaps the prop should be `items` rather than `actionItems`? I don't feel these are particularly action-y anyway, they're just menu items right?

File: plugins/playlist-backend/src/service/router.ts
Comment (Rugvip) [inline plugins/playlist-backend/src/service/router.ts:235]: Perhaps mention here that this enforces catalog permissions as well? feeling this could easily be removed without considers that by mistake right now

**Review by Rugvip — CHANGES_REQUESTED** [review]
Nice :grin:

Just some minor stuff, and waiting for 🟢  from @backstage/techdocs-core regarding the search changes

**Comment by kuangp** [comment]
@Rugvip updated !

**Review by Rugvip — APPROVED** [review]
👍  will try to get a +1 for the search changes tomorrow too so we can :shipit: 🎉

**Comment by kuangp** [comment]
cc: @iamEAP @emmaindal and the techdocs folks if you guys can take a look when you get the chance please 🙏 
Would love to get this get for the next stable release 😁

File: plugins/search-react/src/components/SearchFilter/SearchFilter.Autocomplete.tsx
Comment (camilaibs) [inline plugins/search-react/src/components/SearchFilter/SearchFilter.Autocomplete.tsx:82]: To be sure I understood, this is for requesting the first results page whenever a filter change, right?

Comment (camilaibs) [reply plugins/search-react/src/components/SearchFilter/SearchFilter.Autocomplete.tsx:82]: If so, I believe doing this in the search context would benefit anyone who changes filters on other files, similar to this:
https://github.com/backstage/backstage/blob/535ca54f0e7ccc39eedefe084b94cb8aa42f7df9/plugins/search-react/src/context/SearchContext.tsx#L138-L144

File: plugins/search-react/src/components/SearchFilter/SearchFilter.test.tsx
Comment (camilaibs) [inline plugins/search-react/src/components/SearchFilter/SearchFilter.test.tsx:142]: Could you test clearing the filter in the search context if you agree with the idea?

File: plugins/search-react/src/components/SearchFilter/SearchFilter.tsx
Comment (camilaibs) [reply plugins/search-react/src/components/SearchFilter/SearchFilter.tsx:109]: The same as commented in the `SearchFilter` autocomplete 🙂

**Review by camilaibs — COMMENTED** [review]
Thanks for contributing again, just one suggestion:

**Comment by kuangp** [comment]
@camilaibs updated based on your suggestion, thanks for taking a look!

**Review by camilaibs — APPROVED** [review]
Looks great to me ⭐

**Pull request merged**
